testjson/execution.go

```diff
@@ -538,15 +546,41 @@ func (e *Execution) Failed() []TestCase {
 	return failed
 }
 
-// FilterFailedUnique filters a slice of failed TestCases by removing root test
-// case that have failed subtests.
+// FilterFailedUnique filters a slice of failed TestCases to remove any parent
+// tests that have failed subtests. The parent test will always be run when
+// running any of its subtests.
 func FilterFailedUnique(tcs []TestCase) []TestCase {
-	var result []TestCase
+	sort.Slice(tcs, func(i, j int) bool {
+		a, b := tcs[i], tcs[j]
+		if a.Package != b.Package {
+			return a.Package < b.Package
+		}
+		return len(a.Test.Name()) > len(b.Test.Name())
+	})
+
+	var result []TestCase //nolint:prealloc
+	var parents = make(map[string]map[string]bool)
```

Comment:
Since you just care about the existence in the map, you can use `struct{}` instead of `bool`. Go understands this and it helps reduce memory usage.
```suggestion
	var parents = make(map[string]map[string]struct{})
```

Reply:
I usually would use `struct{}`, but recently a colleague has been trying to convince me that using a `bool` is slightly better because the memory saving of `struct{}` is negligible, and since the use of this `map` is very localized making a mistake of storing a false is unlikely. The readability of bool would then out weight the benefits of a `struct{}`. I thought I'd give it a shot here.

Reply:
> mistake of storing a false is unlikely

Even if you stored a false, it wouldn't matter since you never examine the actual value.

Reply:
This is your code base, so whatever you prefer is fine.